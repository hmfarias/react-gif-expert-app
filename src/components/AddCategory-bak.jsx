import { useState } from 'react';

export const AddCategory = ({ setCategories }) => {
	const [inputValue, setInputValue] = useState('');

	//Puedo recibir el parámetro estructurado:
	// const onInputChange = (event) => {
	// 	setInputValue(event.target.value);
	// };

	//Puedo recibir el parámetro desestructurado:
	const onInputChange = ({ target }) => {
		// console.log(target.value);
		setInputValue(target.value);
	};

	const onSubmit = (event) => {
		event.preventDefault();
		if (inputValue.trim().length < 1) return; //si se da enter en el input sin que haya valores, se retorna sin hacer nada
		setCategories((categories) => [...categories, inputValue]); //agrego lo que se haya escrito a la lista de categorias
		setInputValue(''); //limpio la caja del input
		// console.log(inputValue);
	};

	return (
		// <form onSubmit={(event)=> onSubmit(event)}> puede ser así, o simplificado como abajo
		//onSubmit refresca por defecto toda la pagina. Eso lo evito con event.preventDefault en el onSubmit.
		<form onSubmit={onSubmit}>
			<input
				type="text"
				placeholder="Buscar Gifs"
				value={inputValue}
				// onChange={(event) => onInputChange(event)} //puedo mandarlo así
				onChange={onInputChange} //o simplificarlo así - siempre se va a mandar el event
			/>
		</form>
	);
};
