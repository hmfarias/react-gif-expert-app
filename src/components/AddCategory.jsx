import { useState } from 'react';
import { capitalizeString } from '../helpers';

export const AddCategory = ({ onNewCategory }) => {
	//para mantener el valor ingresado en el input uso el hoock useState
	const [inputValue, setInputValue] = useState(''); //2

	//Puedo recibir el parámetro estructurado:
	// const onInputChange = (event) => {
	// 	setInputValue(event.target.value);
	// };
	//Puedo recibir el parámetro desestructurado:
	const onInputChange = ({ target }) => {
		//3
		// console.log(target.value);
		setInputValue(target.value);
	};

	const onSubmit = (event) => {
		//5
		event.preventDefault(); //para que no haga el full Refresh de la pagina
		const newInputValue = inputValue.trim(); //le saco los espacios al principio o al final
		if (newInputValue.length < 1) return; //si se da enter en el input sin que haya valores, se retorna sin hacer nada

		//pongo la primera letra de cada palabra en mayuscula
		const newImputValueCapital = capitalizeString(newInputValue);

		onNewCategory(newImputValueCapital); //mando la nueva categoria al padre para que la inserte

		setInputValue(''); //limpio la caja del
		// console.log(inputValue);
	};

	return (
		//creo un form para que cuando se le dé enter se dispare el submit de ese form y ahi es donde voy a agregar la categoria
		// <form onSubmit={(event)=> onSubmit(event)}> puede ser así, o simplificado como abajo
		//onSubmit refresca por defecto toda la pagina. Eso lo evito con event.preventDefault en el onSubmit.

		//4
		<form onSubmit={onSubmit}>
			<input //1
				type="text"
				placeholder="Buscar Gifs"
				value={inputValue}
				// onChange={(event) => onInputChange(event)} //puedo mandarlo así
				//para poder cambiar el input, react pide que se agregue el metodo onChange si no no se puede cambiar el valor por mas que escriba
				onChange={onInputChange} //o simplificarlo así - siempre se va a mandar el event
			/>
		</form>
	);
};
