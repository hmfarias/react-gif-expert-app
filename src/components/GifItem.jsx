import PropTypes from 'prop-types';

export const GifItem = ({ title, url }) => {
	return (
		<div className="card">
			<img src={url} alt={title} />
			<p>{title}</p>
		</div>
	);
};

// PROP TYPES
GifItem.propTypes = {
	title: PropTypes.string.isRequired,
	url: PropTypes.string.isRequired,
};

//DEFAULT PROPS
// GifItem.defaultProps = {
// 	value: 0,
// };
