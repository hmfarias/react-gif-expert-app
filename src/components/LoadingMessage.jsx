export const LoadingMessage = ({ state }) => {
	return state ? <h2>Loading ...</h2> : null;
};
