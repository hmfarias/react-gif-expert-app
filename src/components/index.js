//archivo de barril que facilita las importaciones en los distintos componentes

export * from './AddCategory';
export * from './GifGrid';
export * from './GifItem';
