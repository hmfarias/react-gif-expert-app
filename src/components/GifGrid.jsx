// import { useEffect, useState } from 'react'; //ya no lo uso porque paso a useFetchGifs
// import { getGifs } from '../helpers/getGifs';//ya no lo uso porque paso a useFetchGifs
import { GifItem } from './GifItem';
import { useFetchGifs } from '../hooks/useFetchGifs';
import { LoadingMessage } from './LoadingMessage';

export const GifGrid = ({ category }) => {
	const { images, isLoading } = useFetchGifs(category);

	//todo esto de abajo que hace la carga de las imagenes
	//lo voy a llevar a un custom hook useFetchGifs**********
	// //Para mantener las imagenes uso el useState
	// const [images, setImages] = useState([]); //inicializado con array vacio

	// //ahora pongo las imagenes que trajo getGifs en el array images
	// //getImages devuelve una promesa y por ello uso async await
	// //no puedo usar async directamente en el useEffect y por eso creo la funcion getImages por fuera del useEffect
	// const getImages = async () => {
	// 	const newImages = await getGifs(category);
	// 	setImages(newImages);
	// };

	// //useEffect es un hook de react que permite definir cuando queramos que se ejecute un efecto secundario
	// //lleva un call back donde se define el efecto que queremos disparar y como segundo argumento un arreglo donde se pone una lista de dependencias de react que si se deja vacío le indica que sólo se ejecuta la primera vez que se construye el componente
	// useEffect(() => {
	// 	getImages();
	// }, []);
	//todo esto lo de arriba lo voy a remplazar con un custom hook **********

	return (
		<>
			<h3>{category}</h3>
			{
				//puedo hacerlo usando operador ternario
				// isLoading ? <h2>Loading ...</h2> : null

				//o puedo hacerlo usando un if corto con una sola condicion usando el and logico &&
				// isLoading && <h2>Loading ...</h2>
				<LoadingMessage state={isLoading} />
			}
			<div className="card-grid">
				{images.map((img) => (
					<GifItem
						key={img.id}
						{...img} //de este modo envio al componente GifItem todas las propiedades que hay en img y puedo desestructurar la que necesite ya en el componente
					/>
				))}
			</div>
		</>
	);
};
