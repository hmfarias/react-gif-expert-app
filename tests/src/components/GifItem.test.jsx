import { render, screen } from '@testing-library/react';
import { GifItem } from '../../../src/components/GifItem';

describe('<Gifitem /> component tests', () => {
	const title = 'Nesdimug';
	const url = 'http://www.notredamejoyas.com.ar/';

	test('should be the same as Snapshot', () => {
		//inicializacion
		// estímulo
		const { container } = render(<GifItem title={title} url={url} />);
		// console.log(container);
		//observar el comportamiento esperado
		expect(container).toMatchSnapshot();
	});

	test('should show the image with the URL and the ALT indicated', () => {
		render(<GifItem title={title} url={url} />);

		// //Puedo hacerlo así:
		// //primero tomo la imagen u verifico que el url que tenga sea igual que el enviado
		// // console.log(screen.getByRole('img').src);
		// expect(screen.getByRole('img').src).toBe(url);
		// //y lo mismo para el ALT
		// expect(screen.getByRole('img').alt).toBe(title);

		//O mucho mejor hacerlo usando desestructuracion
		const { src, alt } = screen.getByRole('img');
		expect(src).toBe(url);
		expect(alt).toBe(title);
	});

	test('should show title sent ', () => {
		render(<GifItem title={title} url={url} />);
		// screen.debug();
		expect(screen.getByText(title)).toBeTruthy(); //simplemente que exista el texto del titulo
	});
});
