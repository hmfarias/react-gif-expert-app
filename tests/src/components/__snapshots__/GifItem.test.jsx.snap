// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Gifitem /> component tests should be the same as Snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Nesdimug"
      src="http://www.notredamejoyas.com.ar/"
    />
    <p>
      Nesdimug
    </p>
  </div>
</div>
`;
